content/knowledge_base/flare/consensus/avalanche/specification.md

```diff
@@ -0,0 +1,210 @@
+
+generateTx:
+    edges = {T' <— T: T' ∈ parentSelection(`Set T`)}
+    T = Tx(data, edges)
+    onReceiveEvent(T)
+
+onReceiveEvent(t)
+    if T ∉ `Set T` then
+        if PT (conflict set of T) = Ø then
+            PT = {T}, Pt.pref = T
+            PT.last = T, PT.cnt = 0
+        else PT = PT U {T}
+        `Set T` = T U {T}, cT = 0
+```
+
+Because transactions that generate and consume the same UTXO do not conflict with each other, any transaction can be reissued with different parents.
+
+### Node Main Loop
+
+Protocol main loop executed by each node:
+
+```text
+AvalancheLoop:
+    while true do
+        find T that satisfies T ∈ `Set T` AND T ∉ Q
+        K = sample(N\u, k)
+        P = SUM(v ∈ k) query(v, T)
+
+        if P ≥ α then
+            cT = 1 (chit for transaction T)
+            update preferences for ancestors:
+            for T' ∈ `Set T`: T' *<— T do
+                if T' confidence is larger than the confidence of preferred transation T` in the P conflict set - update the prefered transaction of the set:
+                if d(T') > d(PT'.pref) then
+                    PT'.pref = T'
+                if T' is not the last transaction in the T' conflict set - update the last place with it
+                if T' ≠ PT'.last then
+                    PT'.last = T', PT'.cnt=1
+                else
+                    ++PT'.cnt
+        else
+            for T' ∈ `Set T`: T' *<— T do
+                PT'.cnt = 0
+
+            otherwise confidence cnt remains 0 forever
+            Q = Q U {T} - mark T as queried
+```
+
+In each iteration the node tries to select a transaction that has not yet been queried.
+If no such transaction exists, the loop will stall until a new transation is added to the Set T.
+The node then select k number of peers and queries them.
+If more than α of these peers return a positive response, the chit value is set to 1.
+After this tep the loop updates the prefered transaction of each conflict set of the transactions in its ancestry.
+Next, T is added to the set Q of queried transactions so it will never be queried again by the node.
+
+### Query for Transaction
+
+What happens when a node receives a query for transaction T from peer j:
+
+```text
+isPreferred(T)
+    return T = PT.pref
+
+isStronglyPreferred(T)
+    return ∀T' ∈ `Set T`, T' *<— T : isPreferred(T')
+
+isAccepted(T)
+    return
+        ((∀T' ∈ `Set T`, T' <— T: isAccepted(T'))
+            ∩ |PT|=1 ∩ PT.cnt ≥ β1)  safe early commitment
+        ∪ (PT.cnt ≥ β2)              consecutive counter
+
+onQuery(j, T)
+    onReceiveTx(T)
+    respond(, isStronglyPreferred(T))
+```
+
+Firt step is to add transaction T to `Set T` unless it is already included.
+Atfer that step `u` determines in T is currently strongly preferred, and if it is `u` returns a positive response to peer `j`, otherwise returns a negative response.
+In the example pseudocode it is assumed that if a node knows about T, then it recursively knows about T ancestry.
```

Comment:
```suggestion
In the example pseudocode it is assumed that if a node knows about `T`, then it recursively knows about `T`'s ancestry.
```

By recursively knows, do you mean that it needs to do a recursive function call to fetch `T`'s ancestry? I might have missed something here 🤔

Reply:
No, the DAG is used to batch queries for children and parents in order to escape recursive querying.
Here what is meant is that the node knows about each step to next/previous in the ancestry/progeny line.
When I read it now, it makes sense to be updated though

Added the quotation marks here https://github.com/optakt/optakt.github.io/pull/13/commits/feb634318c2fa35f9a26038da3fcf0a5472292ac#diff-2d4f16e70781857aec6689a748ad155e45611b81f184aee076bb139890410b98R190
Thanks!